pkg/azure/blob.go

```diff
@@ -0,0 +1,257 @@
+		}
+	}
+
+	// Extract top level directory.
+	objectDir, _ := filepath.Split(localPath)
+	if objectDir != "" {
+		// Create any missing top level directories.
+		if err := os.MkdirAll(objectDir, 0700); err != nil {
+			return "", err
+		}
+	}
+
+	// Download object.
+	res, err := blob.Download(ctx, nil)
+	if err != nil {
+		return "", err
+	}
+
+	// Prepare target file.
+	f, err := os.OpenFile(localPath, os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return "", err
+	}
+
+	// Calculate hash during write.
+	hash := md5.New()
+
+	// Off we go.
+	mw := io.MultiWriter(f, hash)
+	if _, err = io.Copy(mw, res.Body(nil)); err != nil {
+		if err = f.Close(); err != nil {
+			ctrl.LoggerFrom(ctx).Error(err, "failed to close file after copy error")
+		}
+		return "", err
+	}
+	if err = f.Close(); err != nil {
+		return "", err
+	}
+	return *res.ETag, nil
+}
+
+// VisitObjects iterates over the items in the provided object storage
+// bucket, calling visit for every item.
+// If the underlying client or the visit callback returns an error,
+// it returns early.
+func (c *BlobClient) VisitObjects(ctx context.Context, bucketName string, visit func(path, etag string) error) error {
+	container := c.ServiceClient.NewContainerClient(bucketName)
+
+	items := container.ListBlobsFlat(&azblob.ContainerListBlobFlatSegmentOptions{})
+	for items.NextPage(ctx) {
+		resp := items.PageResponse()
+
+		for _, blob := range resp.ContainerListBlobFlatSegmentResult.Segment.BlobItems {
+			if err := visit(*blob.Name, fmt.Sprintf("%x", *blob.Properties.Etag)); err != nil {
+				err = fmt.Errorf("listing objects from bucket '%s' failed: %w", bucketName, err)
+				return err
+			}
+		}
+	}
+	if err := items.Err(); err != nil {
+		err = fmt.Errorf("listing objects from bucket '%s' failed: %w", bucketName, err)
+		return err
+	}
+	return nil
+}
+
+func (c *BlobClient) Close(_ context.Context) {
+	return
+}
+
+// ObjectIsNotFound checks if the error provided is an azblob.StorageError with
+// an azblob.StorageErrorCodeBlobNotFound error code.
+func (c *BlobClient) ObjectIsNotFound(err error) bool {
+	var stgErr *azblob.StorageError
+	if errors.As(err, &stgErr) {
+		if stgErr.ErrorCode == azblob.StorageErrorCodeBlobNotFound {
+			return true
+		}
+	}
+	return false
+}
+
+// buildServiceClient attempts to build an azblob.ServiceClient with
+// authentication configured based on the data from the Secret.
+//
+// It configures (in order):
+// - azidentity.ManagedIdentityCredential for a Resource ID, when a
+//   "resourceId" entry is found.
+// - azidentity.ManagedIdentityCredential for a User ID, when a "clientId"
+//   but no "tenantId" entry is found.
+// - azidentity.ClientSecretCredential when a "tenantId", "clientId" and
+//   "clientSecret" is found.
+// - azblob.SharedKeyCredential when a "accountName" and "accountKey" are
+//   found.
+// - Client without credentials.
+func buildServiceClient(obj *sourcev1.Bucket, secret *corev1.Secret) (_ azblob.ServiceClient, err error) {
```

Comment:
@laozc can you please take a look and see if this does not cover anything your version did?

Reason for this change is that I had a look at the current Azure libraries out there, and this seems to be the one with the brightest feature and maintenance expectations. Given this is a new dependency introduction, this has my preference over one we eventually need to replace (or semi-copy/borrow from another project).

Reply:
Hi @hiddeco

I'm still verifying this PR in Azure environments.
I think it would be better to use the following order during authentication

1. Check for client secret first to use Service Principal login
2. Use Client ID with Managed Identity login
3. Use Resource ID with Managed Identity login
4. Use Storage Account key based auth login
5. No auth

This may ensure it works as expected when multiple Managed Identites bound on the same VM node for 2) and 3).

```go
func tokenCredentialFromSecret(secret *corev1.Secret) (azcore.TokenCredential, error) {
	tenantID, hasTenantID := secret.Data[tenantIDField]
	clientID, hasClientID := secret.Data[clientIDField]
	clientSecret, hasClientSecret := secret.Data[clientSecretField]
	resourceID, hasResourceID := secret.Data[resourceIDField]

	if hasTenantID && hasClientID && hasClientSecret && string(clientSecret) != "" {
		return azidentity.NewClientSecretCredential(string(tenantID), string(clientID), string(clientSecret), nil)
	}

	if hasClientID {
		return azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
			ID: azidentity.ClientID(clientID)})
	} else if hasResourceID {
		return azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
			ID: azidentity.ResourceID(resourceID)})
	}
	return nil, nil
}
```

Let me get back to you when I have all these scenarios verified.

Reply:
Thanks for the heads up, I will take your suggestions into account and document the reasoning in-code. Please let me know if anything else pops up, and I'll try to address it swiftly.

Reply:
I have updated the code to your suggestion, you may also want to provide some input on https://github.com/fluxcd/source-controller/pull/598#pullrequestreview-899089163 and https://github.com/fluxcd/source-controller/pull/598#issuecomment-1058248274.